h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -906,17 +906,43 @@ protected boolean ignoreUuidColumns() {
    */
   protected void ignoreBadColumns(int npredictors, boolean expensive){
     // Drop all-constant and all-bad columns.
-    if(_parms._ignore_const_cols)
+    if(_parms._ignore_const_cols && _parms._checkpoint == null) {
       new FilterCols(npredictors) {
-        @Override protected boolean filter(Vec v) {
+        @Override
+        protected boolean filter(Vec v) {
           boolean isBad = v.isBad();
           boolean skipConst = ignoreConstColumns() && v.isConst();
           boolean skipString = ignoreStringColumns() && v.isString();
           boolean skipUuid = ignoreUuidColumns() && v.isUUID();
           boolean skip = isBad || skipConst || skipString || skipUuid;
           return skip;
         }
-      }.doIt(_train,"Dropping bad and constant columns: ",expensive);
+      }.doIt(_train, "Dropping bad and constant columns: ", expensive);
+
+
+    } else if(_parms._checkpoint != null){ // If there is a checkpoint, keep exactly the same columns
+      final Model checkpointedModel = _parms._checkpoint.get();
+      assert checkpointedModel != null;
+
+      // Avoid O(n) lookup
+      final Set<String> checkpointNames = new HashSet<>(checkpointedModel._output._names.length);
```

Comment:
For Lookup, a method `find` of `Frame` class could be used. However, the complexity of that method is *O(n)* now. To reduce the comp. complexity of Frame's `find`, a HashSet must be introduced inside Frame. It would hold redudant values to `_names` string array. For many columns, it also means "slightly" bigger memory overhead we have no explicit reason to introduce right now.

This is the reason I did it this way in this place. Quickly instantiate a HashSet with columns listed in the checkpointed model, use it to search few lines later and then drop it. Minimal change right now.

Reply:
Yeap, we are always trading off memory/computation in all kinds of algos.  Feels like the fight of good vs evil going on again.